### PR TITLE
13 - Add routing / 9 - Add base filters

### PR DESCRIPTION
**Pull request by ogavashi**

### CODE REVIEW

File: src/components/Unknown/BuildScreen/Filter/index.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import { Button } from '@mui/material';
+import { Box } from '@mui/system';
+import React, { useCallback, useState } from 'react';
+import useQuery from '../../../../hooks/useQuery';
+import RangeFilter from './RangeFilter';
+
+import useStyles from './styles';
+import SwitchFilter from './SwitchFilter';
+
+const Filters: React.FC = () => {
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/index.tsx]: rename to Filter

File: src/components/Unknown/BuildScreen/Filter/index.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  const styles = useStyles();
+
+  const { handleParamsChange } = useQuery();
+
+  const [filters, setFilters] = useState<Record<string, string>>({});
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/index.tsx]: filters by default - null, not {}

File: src/components/Unknown/BuildScreen/Filter/index.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  const handleChangeFilters = useCallback((filter: Record<string, string>) => {
+    setFilters((prev) => {
+      return {
+        ...prev,
+        ...filter,
+      };
+    });
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/index.tsx]: better to return inline

File: src/components/Unknown/BuildScreen/Filter/index.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  };
+
+  return (
+    <Box className={styles.wrapper}>
+      <SwitchFilter
+        title="cpu"
+        options={[
+          { value: 'INTEL', key: 'intel' },
+          { value: 'AMD', key: 'amd' },
+        ]}
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/index.tsx]: if this is temporary you should add TODO

File: src/components/Unknown/BuildScreen/Filter/styles.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  createStyles({
+    wrapper: {
+      display: 'flex',
+      flexDirection: 'column',
+      alignItems: 'center',
+      padding: theme.spacing(4),
+    },
+    button: {
+      width: '100%',
+    },
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/styles.ts]: Button have fullWidth prop for that

File: src/components/Unknown/NotFoundScreen/styles.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { createStyles, makeStyles } from '@mui/styles';
+import { Theme } from '@mui/system';
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    title: {
+      fontSize: '180px',
```
Comment (VHarastei) [inline src/components/Unknown/NotFoundScreen/styles.ts]: fontSize: 180

File: src/components/Unknown/Root/index.tsx [hunk 7]
```diff
@@ -28,11 +30,17 @@ const Root: React.FC = () => {
...
-    <div>
-      <Layout>
-        <BuildScreen />
-      </Layout>
-    </div>
+    <Layout>
+      <Routes>
+        <Route path="/" element={<BuildScreen />} />
+        <Route path="login" element={<div>Sign In</div>} />
+        <Route path="login" element={<div>Sign Up</div>} />
```
Comment (VHarastei) [inline src/components/Unknown/Root/index.tsx]: path = register

File: src/hooks/usePriceInputs.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
+import { useState } from 'react';
+
+const usePriceInputs = (searchParams: URLSearchParams) => {
+  const [priceRange, setPriceRange] = useState<number[]>([
+    Number(searchParams.get('minPrice')) || 0,
+    Number(searchParams.get('maxPrice')) || 50000,
+  ]);
```
Comment (VHarastei) [inline src/hooks/usePriceInputs.ts]: type PriceRange = {
minPrice: number,
maxPrice: number
}

File: src/hooks/usePriceInputs.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  const minDistance = 1000;
+  const regex = /^[0-9]*$/;
```
Comment (VHarastei) [inline src/hooks/usePriceInputs.ts]: to constants

File: src/hooks/usePriceInputs.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+      if (activeThumb === 0) {
+        const clamped = Math.min(newValue[0], 50000 - minDistance);
+        setPriceRange([clamped, clamped + minDistance]);
+      } else {
+        const clamped = Math.max(newValue[1], minDistance);
+        setPriceRange([clamped - minDistance, clamped]);
+      }
+    } else {
+      setPriceRange(newValue as number[]);
+    }
```
Comment (VHarastei) [inline src/hooks/usePriceInputs.ts]: It is better not to make such nesting of if else blocks
if(activeThumb === 0) ... return

File: src/hooks/usePriceInputs.ts (hunk 9 above)
Comment (VHarastei) [reply src/hooks/usePriceInputs.ts]: regex

File: src/common/constants.ts [hunk 11]
```diff
@@ -8,5 +8,13 @@ const ProductCategoryByCollection = {
   HardDrives: 'HDD',
 };
 
+const NUMERIC_INPUT_FORMAT = /^[0-9]*$/;
+
+const minSliderDistance = 1000;
```
Comment (VasylHarastei) [inline src/common/constants.ts]: Uppercase

File: src/components/Unknown/BuildScreen/Filter/AccordionFilter/index.tsx [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import SwapHorizIcon from '@mui/icons-material/SwapHoriz';
+
+import React, { useContext } from 'react';
+import useStyles from './styles';
+import FilterItem from '../FilterItem';
+import useQuery from '../../../../../hooks/useQuery';
+import useFilterAccordion from '../../../../../hooks/useFilterAccordion';
+import { BuildScreenContext } from '../../../BuildScreenContext';
+
+type Option = {
```
Comment (VasylHarastei) [inline src/components/Unknown/BuildScreen/Filter/AccordionFilter/index.tsx]: FilterOption

File: src/components/Unknown/BuildScreen/Filter/AccordionFilter/index.tsx [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+const AccordionFilter: React.FC<AccordionFilterProps> = ({ filter }) => {
+  const styles = useStyles();
+
+  const { selectedFilters, handleAddFilter } = useFilterAccordion(filter.key);
+
+  const { selectedFilter, handleSelectFilter } = useContext(BuildScreenContext);
+
+  const isSelected = selectedFilters?.length;
```
Comment (VasylHarastei) [inline src/components/Unknown/BuildScreen/Filter/AccordionFilter/index.tsx]: IsSelected now number type, add !!

File: src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx [hunk 14]
```diff
@@ -0,0 +1,45 @@
+import { ButtonBase, Typography } from '@mui/material';
+import { Box } from '@mui/system';
+import React, { useContext } from 'react';
+import { BuildScreenContext } from '../../../BuildScreenContext';
+import useStyles from './styles';
+
+export type Option = {
+  value: string;
+  key: string;
+};
```
Comment (VasylHarastei) [inline src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx]: Duplicate

File: src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+  option,
+  handleAddFilter,
+  selectedFilters,
+}) => {
+  const styles = useStyles();
+  const { handleChangeFilters } = useContext(BuildScreenContext);
+
+  const isSelected = selectedFilters && selectedFilters.includes(option.key);
+
+  const onClickItem = (value: string) => {
```
Comment (VasylHarastei) [inline src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx]: handle, and item is ambiguous

File: src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    const selectedParamsQuery = handleAddFilter(value);
+    handleChangeFilters(selectedParamsQuery);
+  };
+
+  return (
+    <Box className={styles.wrapper}>
+      <ButtonBase
+        className={`${styles.button} ${isSelected && styles.selected}`}
```
Comment (VasylHarastei) [inline src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx]: add clsx, if isSelected false this add false to classname

File: src/components/Unknown/BuildScreen/Filter/FilterItem/styles.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+      borderRadius: theme.shape.borderRadius,
+      margin: theme.spacing(2),
+      cursor: 'pointer',
+    },
+    button: {
+      width: '100%',
+      height: '100%',
+      borderRadius: theme.shape.borderRadius,
+      padding: theme.spacing(2),
+    },
```
Comment (VasylHarastei) [inline src/components/Unknown/BuildScreen/Filter/FilterItem/styles.ts:17]: We don't have button in theme?

File: src/components/Unknown/BuildScreen/Filter/RangeFilter/index.tsx [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+      <Typography gutterBottom variant="h3">
+        {title}:
+      </Typography>
+      <Box display="flex" flexDirection="row" justifyContent="space-between">
+        <TextField
+          className={styles.textField}
+          value={priceRange.minPrice}
+          onChange={handleMinPrice}
+          onBlur={validateRange}
+          inputProps={{ inputMode: 'numeric', pattern: '[0-9]*' }}
```
Comment (VasylHarastei) [inline src/components/Unknown/BuildScreen/Filter/RangeFilter/index.tsx]: Regex to constants

File: src/components/Unknown/BuildScreen/Filter/RangeFilter/index.tsx [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        />
+        <TextField
+          className={styles.textField}
+          value={priceRange.maxPrice}
+          onChange={handleMaxPrice}
+          onBlur={validateRange}
+          inputProps={{ inputMode: 'numeric', pattern: '[0-9]*' }}
+        />
```
Comment (VasylHarastei) [inline src/components/Unknown/BuildScreen/Filter/RangeFilter/index.tsx]: Use inputBase, we have styles in theme

File: src/components/Unknown/NotFoundScreen/index.tsx [hunk 20]
```diff
@@ -1,16 +1,44 @@
...
+        Page not found
+      </Typography>
+      <Button
+        variant="contained"
+        color="primary"
+        size="large"
+        onClick={() => navigate('/')}
+      >
+        Go home
+      </Button>
```
Comment (VasylHarastei) [inline src/components/Unknown/NotFoundScreen/index.tsx]: Prop component={Link}, link from react router

File: src/components/Unknown/Switcher/index.tsx [hunk 21]
```diff
@@ -1,39 +1,46 @@
...
-
-export default Switcher;
+import { ToggleButton, ToggleButtonGroup } from '@mui/material';
+import React from 'react';
+import useStyles from './styles';
+
+export type Option = {
+  key: string;
+  value: string;
+};
```
Comment (VasylHarastei) [inline src/components/Unknown/Switcher/index.tsx]: Duplicate

File: src/components/Unknown/BuildScreen/Filter/FilterItem/styles.ts (hunk 17 above)
Comment (ogavashi) [reply src/components/Unknown/BuildScreen/Filter/FilterItem/styles.ts:17]: This is a customized base button, changing it in theme will lead to distortion of all other buttons

File: src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+  const handleClickFilterItem = (value: string) => {
+    const selectedParamsQuery = handleAddFilter(value);
+    handleChangeFilters(selectedParamsQuery);
+  };
+
+  const buttonClasses = clsx({
+    [styles.button]: true,
+    [styles.selected]: isSelected,
+  });
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx]: there is no need to store it in a variable

Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/FilterItem/index.tsx]: [styles.button]: true looks wierd

File: src/components/Unknown/BuildScreen/Filter/RangeFilter/styles.ts [hunk 23]
```diff
@@ -0,0 +1,15 @@
+import { Theme } from '@mui/material';
+import { createStyles, makeStyles } from '@mui/styles';
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    wrapper: {
+      display: 'flex',
+      flexDirection: 'column',
+      alignItems: 'flex-start',
+      justifyContent: 'center',
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/RangeFilter/styles.ts]: Box doesn`t have this props?

File: src/components/Unknown/BuildScreen/Filter/RangeFilter/index.tsx [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+    validateRange,
+    handleSliderChange,
+  } = usePriceInputs();
+
+  const debouncedValue = useDebounce<PriceRange>(priceRange, 30);
+
+  useEffect(() => {
+    handleChangeFilters({
+      minPrice: String(debouncedValue.minPrice),
+      maxPrice: String(debouncedValue.maxPrice),
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/RangeFilter/index.tsx:35]: why price is string?

File: src/components/Unknown/BuildScreen/Filter/index.tsx [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+
+// const CPUFilters = [
+//   { value: 'INTEL', key: 'intel' },
+//   { value: 'AMD', key: 'amd' },
+// ];
+
+// const GPUFilters = [
+//   { value: 'NVIDIA', key: 'nvidia' },
+//   { value: 'AMD', key: 'amd' },
+// ];
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/index.tsx]: ?

File: src/components/Unknown/BuildScreen/Filter/index.tsx [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+const ThreadsFilter = {
+  title: 'Threads Number',
+  key: 'threads',
+  options: [
+    { value: '4 threads', key: '4 threads' },
+    { value: '8 threads', key: '8 threads' },
+    { value: '12 threads', key: '12 threads' },
+    { value: '16 threads', key: '16 threads' },
+  ],
+};
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/index.tsx]: move this to separate file (in this folder)

File: src/components/Unknown/BuildScreen/Filter/index.tsx [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+        title="cpu"
+        options={CPUFilters}
+      />
+      <SwitchFilter
+        title="gpu"
+        options={GPUFilters}
+      /> */}
+      <RangeFilter title="Budget" />
+      <AccordionFilter filter={OperatingFrequencyFilter} />
+      <AccordionFilter filter={ThreadsFilter} />
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/index.tsx]: but this filters is only for cpu, what about other products? how we handle filters for different products?

File: src/hooks/usePriceInputs.ts [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+        minPrice: atBeginning ? clamped : clamped - MIN_PRICE_SLIDER_DISTANCE,
+        maxPrice: atBeginning ? clamped + MIN_PRICE_SLIDER_DISTANCE : clamped,
+      });
+    } else {
+      setPriceRange({
+        minPrice: newValue[0],
+        maxPrice: newValue[1],
+      });
+    }
+  };
```
Comment (VHarastei) [inline src/hooks/usePriceInputs.ts]:
```
if (newValue[1] - newValue[0] < MIN_PRICE_SLIDER_DISTANCE) {
....
return
}

 setPriceRange({
        minPrice: newValue[0],
        maxPrice: newValue[1],
      });

```

File: src/components/Unknown/BuildScreen/Filter/filters.ts [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+  title: string;
+  key: string;
+  options: FilterOption[];
+};
+
+export type Filters = {
+  [key in Builder]: { accordion?: Filter[]; switcher?: Filter[] };
+};
+
+export const filters: Filters = {
```
Comment (VHarastei) [inline src/components/Unknown/BuildScreen/Filter/filters.ts]: filters isn`t an array, filterByCategory for example (this is not the best name)

File: types/index.ts [hunk 30]
```diff
@@ -137,3 +137,6 @@ export type HardDrive = Product & {
   size: string; // e.g. 147x102x20 mm
   weight: string;
 };
+
+// Product builders
+export type Builder = 'CPU' | 'Graphic card' | 'Motherboard' | 'RAM' | 'HDD';
```
Comment (VHarastei) [inline types/index.ts]: maybe BuilderCategory or smth else, Builder is ambigous

**Review by VHarastei — APPROVED** [review]

**Pull request merged**